syft/rekor/query.go

```diff
@@ -0,0 +1,211 @@
+	}
+	return res.Payload, nil
+}
+
+// Precondition: Client is not nil
+//
+// Postcondition: Rekor's signature over the entry is valid and the inclusion proof that Rekor provides us is valid
+func getAndVerifyRekorEntry(uuid string, client *client.Rekor) (*models.LogEntryAnon, error) {
+	params := entries.NewGetLogEntryByUUIDParams().WithEntryUUID(uuid)
+	res, err := client.Entries.GetLogEntryByUUID(params)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, errors.New("result of rekor query is nil")
+	}
+
+	logEntry := res.Payload
+	if len(logEntry) == 0 {
+		return nil, fmt.Errorf("retrieved rekor entry has no logEntryAnons")
+	}
+	if len(logEntry) > 1 {
+		return nil, fmt.Errorf("retrieved rekor entry has more than one logEntry")
+	}
+
+	// logEntry is a map from uuids to logEntryAnons
+	var logEntryAnon *models.LogEntryAnon
+	for k := range logEntry {
+		val := logEntry[k]
+		logEntryAnon = &val
+	}
+
+	if logEntryAnon.LogIndex == nil {
+		return nil, fmt.Errorf("retrieved rekor entry has no log index")
+	}
+	logIndex := *logEntryAnon.LogIndex
+	log.Debugf("rekor entry %v was retrieved", logIndex)
+
+	ctx := context.Background()
+	if err = cosign.VerifyTLogEntry(ctx, client, logEntryAnon); err != nil {
+		return nil, fmt.Errorf("could not prove that the log entry is on rekor: %w", err)
+	}
+
+	return logEntryAnon, nil
+}
+
+// Precondition: client and its fields are not nil.
+//
+// Postcondition: if no error is returned, the returned struct and its fields are not nil
+func getAndVerifySbomFromUUID(uuid string, client *Client) (*sbomWithMetadata, error) {
+	logEntryAnon, err := getAndVerifyRekorEntry(uuid, client.rekorClient)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving rekor entry by uuid %v: \n\t\t%w", uuid, err)
+	}
+
+	logIndex := *logEntryAnon.LogIndex
+
+	if err = verify(client.rekorClient, logEntryAnon); err != nil {
+		return nil, fmt.Errorf("rekor log entry %v could not be verified: \n\t\t%w", logIndex, err)
+	}
+
+	log.Debugf("verification of rekor entry %v complete", logIndex)
+
+	att, err := parseAndValidateAttestation(logEntryAnon)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing or validating attestation associated with rekor entry %v: \n\t\t%w", logIndex, err)
+	}
+
+	sbomBytes, err := getSbom(att, client.httpClient)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving sbom from rekor entry %v: \n\t\t%w", logIndex, err)
+	}
+
+	log.Debugf("SBOM (%v bytes) retrieved", len(*sbomBytes))
+
+	if err = verifySbomHash(att, sbomBytes); err != nil {
+		return nil, fmt.Errorf("could not verify retrieved sbom (from rekor entry %v): \n\t\t%w", logIndex, err)
+	}
+
+	sbomSha1 := sha1.Sum(*sbomBytes) //nolint:gosec // sha1 needed for checksums in spdx format
+	decodedHash := hex.EncodeToString(sbomSha1[:])
+
+	sbom, err := parseSbom(sbomBytes)
+	if err != nil || sbom == nil {
+		return nil, fmt.Errorf("error parsing sbom from rekor entry %v: %w", logIndex, err)
```

Comment:
the two error types should be handled differently since if `sbom==nil` err is nil. If using `errors` for wrapping this would not be considered an error. This should be a different error message instead.

Reply:
Re second comment: If the retrieved sbom bytes are empty, then I believe that parseSbom will return nil. I added a post condition to getSbom that the returned byte slice is not empty.